packages/common/src/controllers/AccessController.ts

```diff
@@ -62,23 +61,27 @@ export default class AccessController {
       return;
     }
 
-    try {
-      const accessTokens = await this.generateOrRefreshAccessTokens();
+    const getMediaWithPassport = async (forceGenerate: boolean = false) => {
+      const accessTokens = forceGenerate ? await this.generateAccessTokens() : await this.generateOrRefreshAccessTokens();
       if (!accessTokens?.passport) {
-        throw new Error('Failed to get / generate access tokens and retrieve media.');
+        throw new Error('Failed to generate / refresh access tokens.');
       }
-      return await this.apiService.getMediaByIdWithPassport({ id: mediaId, siteId: this.siteId, planId: entitledPlan.id, passport: accessTokens.passport });
+
+      return await this.apiService.getMediaByIdWithPassport({
+        id: mediaId,
+        siteId: this.siteId,
+        planId: entitledPlan.id,
+        passport: accessTokens.passport,
+      });
+    };
+
+    try {
+      return await getMediaWithPassport();
     } catch (error: unknown) {
-      if (error instanceof ApiError && error.code === 403) {
-        // If the passport is invalid or expired, refresh the access tokens and try to get the media again.
-        const accessTokens = await this.refreshAccessTokens();
-        if (accessTokens?.passport) {
-          return await this.apiService.getMediaByIdWithPassport({ id: mediaId, siteId: this.siteId, planId: entitledPlan.id, passport: accessTokens.passport });
-        }
-
-        throw new Error('Failed to refresh access tokens and retrieve media.');
-      }
-      throw error;
+      // If the initial attempt fails, it may indicate the passport does not have the latest plans.
+      // Force new access tokens generation and retry fetching the media.
+      // TODO: Revisit this logic once we address the passport update possibility in Q4
```

Comment:
@kiremitrov123 I would appreciate if you mention all the possible improvements marked as TODO in a separate ticket, so that we had more visibility here.

Reply:
We don't have tickets yet, this will all be a part of the new spotlight. But as soon as we create them I can provide an update with the corresponding tickets.